tests/MongoDB.Driver.Tests/Specifications/unified-test-format/UnifiedValueMatcher.cs

```diff
@@ -0,0 +1,214 @@
+﻿/* Copyright 2020-present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using FluentAssertions;
+using MongoDB.Bson;
+
+namespace MongoDB.Driver.Tests.Specifications.unified_test_format
+{
+    public class UnifiedValueMatcher
+    {
+        private EntityMap _entityMap;
+
+        public UnifiedValueMatcher(EntityMap entityMap)
+        {
+            _entityMap = entityMap;
+        }
+
+        public void AssertValuesMatch(BsonValue expected, BsonValue actual)
+        {
+            AssertValuesMatch(expected, actual, true);
+        }
+
+        private void AssertValuesMatch(BsonValue expected, BsonValue actual, bool isRoot)
+        {
+            if (expected.IsBsonDocument &&
+                expected.AsBsonDocument.ElementCount == 1 &&
+                expected.AsBsonDocument.GetElement(0).Name.StartsWith("$$"))
+            {
+                var expectedDocument = expected.AsBsonDocument; // TODO: Recheck if non document allowed
+                switch (expectedDocument.GetElement(0).Name)
+                {
+                    case "$$unsetOrMatches":
+                        // TODO: recheck
+                        if (actual == null)
+                        {
+                            return;
+                        }
+                        expected = expectedDocument[0];
+                        break;
+                    case "$$type":
+                        AssertExpectedType(actual, expectedDocument[0]);
+                        return;
+                    case "$$matchesHexBytes":
+                        // TODO: recheck
+                        expected = expectedDocument[0];
+                        break;
+                    default:
+                        throw new NotSupportedException($"Special operator not supported: '{expectedDocument.GetElement(0).Name}'");
+                }
+            }
+
+            if (expected.IsBsonDocument)
+            {
+                actual.IsBsonDocument.Should().BeTrue($"Actual value must be a document, but is '{actual.BsonType}'");
```

Comment:
Line 69 can be simply:

    actual.BsonType.Should().Be(BsonType.Document);

Reply:
Great suggestion!
Done